scrapy_po/utils.py

```diff
@@ -1,7 +1,122 @@
 # -*- coding: utf-8 -*-
+import inspect
+from typing import Tuple, Dict, Type, Callable
+
+import andi
+from scrapy.http import Response
+from scrapy.utils.defer import maybeDeferred_coro
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from scrapy_po.webpage import Injectable
+from scrapy_po.page_input_providers import providers
+
 
 def get_callback(request, spider):
-    """ Get request.callback of a scrapy.Request, as a callable """
+    """Get request.callback of a scrapy.Request, as a callable."""
     if request.callback is None:
         return getattr(spider, 'parse')
     return request.callback
+
+
+class DummyResponse(Response):
+
+    def __init__(self, url, request=None):
+        super(DummyResponse, self).__init__(url=url, request=request)
+
+
+def is_callback_using_response(callback: Callable):
+    """Check whether the request's callback method is going to use response."""
+    spec = inspect.getfullargspec(callback)
+    try:
+        arg_name = spec.args[1]  # first index is self, second is response
+    except IndexError:
+        # Parse method is probably using *args and **kwargs annotation.
+        # Let's assume response is going to be used.
+        return True
+
+    if arg_name not in spec.annotations:
+        # There's no type annotation, so we're probably using response here.
+        return True
+
+    if issubclass(spec.annotations[arg_name], DummyResponse):
+        # Type annotation is DummyResponse, so we're probably NOT using it.
+        return False
+
+    # Type annotation is not DummyResponse, so we're probably using it.
+    return True
+
+
+def are_dependencies_using_response(plan: andi.Plan):
+    """Check whether any injectable provider makes use of a valid Response."""
+    for obj, _ in plan:
+        provider = providers.get(obj)
+        if not provider:
+            # Provider not found.
+            continue
+
+        spec = inspect.getfullargspec(provider)
+        for cls in spec.annotations.values():
+            if not issubclass(cls, Response):
+                # Type annotation is not a sub-class of Response.
+                continue
+
+            if issubclass(cls, DummyResponse):
+                # Type annotation is a DummyResponse.
+                continue
+
+            # Type annotation is a sub-class of Response, but not a sub-class
+            # of DummyResponse, so we're probably using it.
+            return True
+
+    # Could not find any Response type annotation in the used providers.
+    return False
+
+
+def is_response_going_to_be_used(request, spider):
+    """Check whether the request's response is going to be used."""
+    callback = get_callback(request, spider)
+    plan, _ = build_plan(callback, {})
+
+    if is_callback_using_response(callback):
```

Comment:
it seems this check can be done before creating a plan, so that you don't need to call build_plan if it passes

Reply:
Done.